src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs

```diff
@@ -248,4 +249,40 @@ public async Task MapAction_HandlesCompletedTaskReturn()
         await endpoints[1].RequestDelegate(httpContext);
         await VerifyResponseBodyAsync(httpContext, string.Empty);
     }
+
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public async Task RequestDelegateWritesAsJsonResponseBody_UnspeakableType(bool useJsonContext)
+    {
+        var source = """
+app.MapGet("/todos", () => GetTodosAsync());
+
+static async IAsyncEnumerable<JsonTodo> GetTodosAsync()
+{
+    yield return new JsonTodo() { Id = 1, IsComplete = true, Name = "One" };
+
+    // ensure this is async
+    await Task.Yield();
+
+    yield return new JsonTodoChild() { Id = 2, IsComplete = false, Name = "Two", Child = "TwoChild" };
+}
+""";
+        var (_, compilation) = await RunGeneratorAsync(source);
+        var serviceProvider = CreateServiceProvider(serviceCollection =>
+        {
+            if (useJsonContext)
+            {
+                serviceCollection.ConfigureHttpJsonOptions(o => o.SerializerOptions.TypeInfoResolver = SharedTestJsonContext.Default);
+            }
+        });
+        var endpoint = GetEndpointFromCompilation(compilation, serviceProvider: serviceProvider);
+
+        var httpContext = CreateHttpContext(serviceProvider);
+
+        await endpoint.RequestDelegate(httpContext);
+
+        var expectedBody = """[{"id":1,"name":"One","isComplete":true},{"$type":"JsonTodoChild","child":"TwoChild","id":2,"name":"Two","isComplete":false}]""";
+        await VerifyResponseBodyAsync(httpContext, expectedBody);
+    }
```

Comment:
Time permitting, it would be good to move this test to the shared set: (https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs#L2049-L2074).

But we can do this in a follow-up PR as well.

Reply:
I can try fitting it in here to help with the test coverage.

Reply:
Done. Please take a look.